codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ConstrainedCollectionGenerator.kt

```diff
@@ -0,0 +1,229 @@
+            "From" to RuntimeType.From,
+            "TryFrom" to RuntimeType.TryFrom,
+            "ConstraintViolation" to constraintViolation,
+        )
+
+        writer.documentShape(shape, model, note = rustDocsNote(name))
+        constrainedTypeMetadata.render(writer)
+        writer.rustTemplate(
+            """
+            struct $name(pub(crate) $inner);
+            """,
+            *codegenScope,
+        )
+        if (constrainedTypeVisibility == Visibility.PUBCRATE) {
+            Attribute.AllowUnused.render(writer)
+        }
+
+        writer.rustTemplate(
+            """
+            impl $name {
+                /// ${rustDocsInnerMethod(inner)}
+                pub fn inner(&self) -> &$inner {
+                    &self.0
+                }
+
+                /// ${rustDocsIntoInnerMethod(inner)}
+                pub fn into_inner(self) -> $inner {
+                    self.0
+                }
+
+                #{ValidationFunctions:W}
+            }
+
+            impl #{TryFrom}<$inner> for $name {
+                type Error = #{ConstraintViolation};
+
+                /// ${rustDocsTryFromMethod(name, inner)}
+                fn try_from(value: $inner) -> Result<Self, Self::Error> {
+                    #{ConstraintChecks:W}
+
+                    Ok(Self(value))
+                }
+            }
+
+            impl #{From}<$name> for $inner {
+                fn from(value: $name) -> Self {
+                    value.into_inner()
+                }
+            }
+            """,
+            *codegenScope,
+            "ConstraintChecks" to constraintsInfo.map { it.tryFromCheck }.join("\n"),
+            "ValidationFunctions" to constraintsInfo.map { it.validationFunctionDefinition(constraintViolation, inner) }.join("\n"),
+        )
+
+        if (!publicConstrainedTypes && isValueConstrained(shape, model, symbolProvider)) {
+            writer.rustTemplate(
+                """
+                impl #{From}<$name> for #{FullyUnconstrainedSymbol} {
+                    fn from(value: $name) -> Self {
+                        value
+                            .into_inner()
+                            .into_iter()
+                            .map(|v| v.into())
+                            .collect()
+                    }
+                }
+                """,
+                *codegenScope,
+                "FullyUnconstrainedSymbol" to symbolProvider.toSymbol(shape),
+            )
+        }
+
+        if (unconstrainedSymbol != null) {
+            writer.rustTemplate(
+                """
+                impl #{ConstrainedTrait} for $name {
+                    type Unconstrained = #{UnconstrainedSymbol};
+                }
+                """,
+                "ConstrainedTrait" to RuntimeType.ConstrainedTrait(),
+                "UnconstrainedSymbol" to unconstrainedSymbol,
+            )
+        }
+    }
+}
+
+internal sealed class CollectionTraitInfo {
```

Comment:
I like this abstraction and the cleanliness of `fromShape`, kudos. I foresee that after we implement all constraint traits, we'll be able to see the forest for the trees and generalize the different generators further.